File: packages/client-python/src/rocketride/core/runtime/downloader.py
"""
Runtime binary downloader.

Downloads release assets from GitHub and extracts them to
~/.rocketride/runtimes/{version}/.
"""

import os
import stat
import sys
import tarfile
import tempfile
import zipfile
from pathlib import Path
from typing import Callable, Optional

import aiohttp

from ..exceptions import RuntimeNotFoundError
from .paths import runtimes_dir, runtime_binary, ensure_dirs
from .platform import asset_name, release_tag

_GITHUB_DOWNLOAD = 'https://github.com/rocketride-org/rocketride-server/releases/download'


async def download_runtime(
    version: str,
    on_progress: Optional[Callable[[int, int], None]] = None,
    on_phase: Optional[Callable[[str], None]] = None,
) -> Path:
    """Download and extract the runtime binary for the given version.

    Returns the path to the runtime binary. No-ops if the binary already exists.

    *on_progress(downloaded_bytes, total_bytes)* is called after each chunk.
    *total_bytes* may be 0 when the server omits Content-Length.

    *on_phase(phase)* is called at each lifecycle transition:
    ``'downloading'``, ``'extracting'``, ``'done'``.
    """
    binary = runtime_binary(version)
    if binary.exists():
        return binary

    ensure_dirs()

    asset = asset_name(version)
    tag = release_tag(version)
    url = f'{_GITHUB_DOWNLOAD}/{tag}/{asset}'

    # Download to a temp file first
    tmp_fd, tmp_path = tempfile.mkstemp(suffix='.download')
    try:
        async with aiohttp.ClientSession() as session:
            async with session.get(url) as resp:
                if resp.status != 200:
                    raise RuntimeNotFoundError(f'Failed to download runtime v{version}: HTTP {resp.status} from {url}')
                total_bytes = resp.content_length or 0
                downloaded_bytes = 0
                if on_phase:
                    on_phase('downloading')
                with os.fdopen(tmp_fd, 'wb') as f:
                    async for chunk in resp.content.iter_chunked(8192):
                        f.write(chunk)
                        downloaded_bytes += len(chunk)
                        if on_progress:
                            on_progress(downloaded_bytes, total_bytes)
                # fd is now closed, don't close again in finally
                tmp_fd = -1

        # Extract
        if on_phase:
            on_phase('extracting')
        dest = runtimes_dir(version)
        dest.mkdir(parents=True, exist_ok=True)

        if asset.endswith('.tar.gz'):
            with tarfile.open(tmp_path, 'r:gz') as tar:
                tar.extractall(path=str(dest))
        elif asset.endswith('.zip'):
            with zipfile.ZipFile(tmp_path, 'r') as zf:
                zf.extractall(path=str(dest))

        # Set executable permission on Unix
        if sys.platform != 'win32' and binary.exists():
            binary.chmod(binary.stat().st_mode | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH)

        if not binary.exists():
            # The archive might have a different internal structure.
            # Look for any executable in the extracted directory.
            raise RuntimeNotFoundError(f'Downloaded and extracted v{version} but binary not found at {binary}. Check the release asset structure.')

        if on_phase:
            on_phase('done')

        return binary

    finally:
        # Clean up temp file
        if tmp_fd >= 0:
            os.close(tmp_fd)
        try:
            os.unlink(tmp_path)
        except OSError:
            pass

File: packages/client-typescript/src/client/runtime/downloader.ts
/**
 * Runtime binary downloader.
 *
 * Downloads release assets from GitHub and extracts them to
 * ~/.rocketride/runtimes/{version}/.
 */

import { createWriteStream, existsSync, chmodSync, mkdirSync, unlinkSync } from 'fs';
import { tmpdir } from 'os';
import { join } from 'path';
import { pipeline } from 'stream/promises';
import { Readable } from 'stream';
import { randomBytes } from 'crypto';
import { RuntimeNotFoundError } from '../exceptions/index.js';
import { runtimesDir, runtimeBinary, ensureDirs } from './paths.js';
import { assetName, releaseTag } from './platform.js';

const GITHUB_DOWNLOAD = 'https://github.com/rocketride-org/rocketride-server/releases/download';

export interface DownloadCallbacks {
	onProgress?: (downloaded: number, total: number) => void;
	onPhase?: (phase: 'downloading' | 'extracting' | 'done') => void;
}

export async function downloadRuntime(version: string, callbacks?: DownloadCallbacks): Promise<string> {
	const binary = runtimeBinary(version);
	if (existsSync(binary)) return binary;

	ensureDirs();

	const asset = assetName(version);
	const tag = releaseTag(version);
	const url = `${GITHUB_DOWNLOAD}/${tag}/${asset}`;

	const ext = asset.endsWith('.zip') ? '.zip' : '.tar.gz';
	const tmpPath = join(tmpdir(), `rr-${randomBytes(8).toString('hex')}${ext}`);

	try {
		// Download
		const resp = await fetch(url, { redirect: 'follow' });

		if (resp.status === 404) {
			throw new RuntimeNotFoundError(`Runtime v${version} not found.\n` + `  No release asset matching "${asset}" exists under GitHub release "${tag}".\n` + `  This usually means the version does not exist or has no binary for your platform.\n` + `  Run "rocketride runtime install latest" to install the latest stable release,\n` + `  or check available versions at:\n` + `  https://github.com/rocketride-org/rocketride-server/releases`);
		}
		if (!resp.ok) {
			throw new RuntimeNotFoundError(`Failed to download runtime v${version}: HTTP ${resp.status} from ${url}`);
		}

		const total = parseInt(resp.headers.get('content-length') ?? '0', 10) || 0;
		callbacks?.onPhase?.('downloading');

		// Stream to temp file with progress tracking
		let downloaded = 0;
		const body = resp.body;
		if (!body) throw new RuntimeNotFoundError('Empty response body');

		const reader = body.getReader();
		let lastProgressTime = 0;
		const nodeStream = new Readable({
			async read() {
				const { done, value } = await reader.read();
				if (done) {
					callbacks?.onProgress?.(downloaded, total);
					this.push(null);
					return;
				}
				downloaded += value.byteLength;
				const now = Date.now();
				if (now - lastProgressTime >= 100) {
					lastProgressTime = now;
					callbacks?.onProgress?.(downloaded, total);
				}
				this.push(Buffer.from(value));
			},
		});

		await pipeline(nodeStream, createWriteStream(tmpPath));

		// Extract
		callbacks?.onPhase?.('extracting');
		const dest = runtimesDir(version);
		mkdirSync(dest, { recursive: true });

		if (asset.endsWith('.tar.gz')) {
			const tar = await import('tar');
			await tar.extract({ file: tmpPath, cwd: dest });
		} else if (asset.endsWith('.zip')) {
			// Use PowerShell on Windows for zip extraction
			const { execSync } = await import('child_process');
			execSync(`powershell -NoProfile -Command "Expand-Archive -Path '${tmpPath}' -DestinationPath '${dest}' -Force"`, { windowsHide: true });
		}

		// Set executable on Unix
		if (process.platform !== 'win32' && existsSync(binary)) {
			chmodSync(binary, 0o755);
		}

		if (!existsSync(binary)) {
			throw new RuntimeNotFoundError(`Downloaded and extracted v${version} but binary not found at ${binary}. Check the release asset structure.`);
		}

		callbacks?.onPhase?.('done');
		return binary;
	} finally {
		try {
			unlinkSync(tmpPath);
		} catch {
			/* already cleaned or never created */
		}
	}
}
